website/src/docs/tus.md

```diff
@@ -87,9 +87,29 @@ When uploading a chunk fails, automatically try again after the millisecond inte
 
 Set to `null` to disable automatic retries, and fail instantly if any chunk fails to upload.
 
-### `retryStatusCodes: [429]`
+### `onShouldRetry: (err, next) => next(err)`
 
-Tus can retry with [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) when receiving 429 (Too Many Requests), or other configured status codes. This can be useful to retry when you hit an API rate limit or when your authentication token is expired.
+When an upload fails `onShouldRetry` is called with the error and the default retry logic as the second argument. The default retry logic is an [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) algorithm triggered on HTTP 429 (Too Many Requests) errors. Meaning if your server (or proxy) returns HTTP 429 because it’s being overloaded, @uppy/tus will find the ideal sweet spot to keep uploading without overloading.
+
+If you want to extend this functionality, for instance to retry on unauthorized requests (to retrieve a new authentication token):
+
+```js
+import Uppy from '@uppy/core'
+import Tus from '@uppy/tus'
+
+const uppy = new Uppy().use(Tus, { endpoint: '', onShouldRetry })
+
+async function onShouldRetry (err) {
+  if (err?.originalResponse?.getStatus() === 401) {
+    return true
+  }
```

Comment:
I really don't think this one makes much sense without the surrounding context, so maybe we can hint that:
```suggestion
  if (err?.originalResponse?.getStatus() === 401) {
    makeSomeAPICallToRequestAuthorization()
    return true
  }
```

Reply:
So I have been thinking about this before commiting but only making a call isn't an example. We would also need to show how you actually update the header. We have an open issue for that and the current workaround is weird: https://github.com/transloadit/uppy/issues/3434#issuecomment-1019341928. I didn't want to include this confusing snippet so that's why I settles on this overly simple example. 

I do agree a proper example would be good.

Reply:
@vp-aridhia could you show how you did it perhaps?

Reply:
The way we're using it is by adding the auth header `onBeforeRequest` and refreshing the token if necessary (on 401s) in the `onAfterResponse`. The underlying issue is with the `onShouldRetry` since it won't retry 401.

I don't really think there's too much gain from having promises on `onShouldRetry` - from a practical perspective, since the API provides `onBeforeRequest` and `onAfterResponse` both support promises (after https://github.com/transloadit/uppy/pull/3712) with which you can pretty much achieve any sort of behaviour you need.

Reply:
Maybe the example could look like this:

```js
import Uppy from '@uppy/core'
import Tus from '@uppy/tus'

let shouldRefreshToken = false
const uppy = new Uppy().use(Tus, {
  endpoint: '',
  onShouldRetry(err, next) {
    if(err?.originalResponse?.getStatus() === 401) {
      shouldRefreshToken = true
      return true
    }
    return next(err)
  },
  async onBeforeRequest() {
    if(shouldRefreshToken) {
      shouldRefreshToken = false
      await makeSomeAPICallToRefreshToken()
    }
  },
})
```

Reply:
Here's how we do it:

```js
const addAuthorizationHeader = async (request) => {
  const token = await authProvider().getIdToken();

  request.setHeader('Authorization', `Bearer ${token}`);
};

const refreshAuthToken = async () => {
  await authProvider().refresh();
};

export const useUppy = () => {
  return useUppyInternal(() => {
    const uppyInstance = new Uppy({
      ...uppyOptions,
    })
      .use(Tus, {
        ...tusOptions,

        onBeforeRequest: async (request) => addAuthorizationHeader(request),
        onAfterResponse: async (_request, response) => {
          const status = response.getStatus();

          if (status === 401) {
            await refreshAuthToken();
          }
        },
      });
  });
};

// from our custom tus plugin, we added in the 401 bit so it retries:

uploadOptions.onShouldRetry = (err) => {
  ...
  } else if (status > 400 && status < 500 && status !== 409 && status !== 401) {
    return false;
  ...
  return true;
};
```

Reply:
Thanks, I improved the example!